feat: bundler module resolution

rin-st opened the pull request:
## Description

Change ts `moduleResolution` to `Bundler`

Context: https://github.com/scaffold-eth/create-eth/pull/74#discussion_r1675316747

Also added paths for abitype/viem/wagmi since with `Bundler` `abi.d.ts` is ignored

Note: Could you check is everything ok with types? My Vscode getting mad, it can show or not show errors for same code. (as you see build works fine)

## Additional Information

- [x] I have read the [contributing docs](/scaffold-eth/scaffold-eth-2/blob/main/CONTRIBUTING.md) (if this is your first contribution)
- [x] This is not a duplicate of any [existing pull request](https://github.com/scaffold-eth/scaffold-eth-2/pulls)

technophile-04 commented:
Thanks @rin-st ! 

> Note: Could you check is everything ok with types? My Vscode getting mad, it can show or not show errors for same code. (as you see build works fine)

Yup it seems to work fine for me in both vscode and neovim. 

But when I update the typescript to 5.5.3, it seems to break again :( 

To reproduce: 

1. Up typescript version

```shell
yarn workspace @se-2/nextjs up typescript
```

2. Check for types: 

```shell
yarn next:check-types
```

Umm, we could merge this already but I am bit skeptical because above thing and also on this branch when I hover over "viem/node_modules/abitype" in `abi.d.ts` it doesn't seem to resolve the path: 

| this branch | main branch | 
| ------- | ----- | 
| ![Screenshot 2024-07-15 at 11 51 26 AM](https://github.com/user-attachments/assets/048abded-a49b-469e-9e16-003caf92feeb) | ![Screenshot 2024-07-15 at 11 52 29 AM](https://github.com/user-attachments/assets/051c49bb-cdcc-403f-bc54-a1d74133ec46) | 

See how on `main` it nicely resolves the whole path

rin-st commented:
I don't know why, but extending `interface Register` doesn't work for me with `moduleResolution: "Bundler"`. 

Possible solution is to rewrite `Address` type directly, not through `Register`. It works, but am I missing something? Do we really need to use `Register`. Also updated TS version and looks like everything works fine. [1528a71](https://github.com/scaffold-eth/scaffold-eth-2/pull/885/commits/1528a711ff2fd6edbd4344f5250baf6ca2e38884) . Of course I'll revert it if something is wrong

As I remember we added `string` because it will be easier to work with our components to newcomers, am I right? Another solution is to remove `abi.d.ts` and add some `as `0x${string}`, but we need to think if it really helps or not

technophile-04 commented on packages/nextjs/tsconfig.json:
```diff
@@ -9,13 +9,16 @@
-      "~~/*": ["./*"]
+      "~~/*": ["./*"],
+      "abitype": ["./node_modules/abitype"],
+      "viem/node_modules/abitype": ["./node_modules/viem/node_modules/abitype"],
+      "wagmi/node_modules/abitype": ["./node_modules/wagmi/node_modules/abitype"]
```
I think we can remove viem and wagmi mapping, from here now?

rin-st replied:
Thanks! [6023ccb](https://github.com/scaffold-eth/scaffold-eth-2/pull/885/commits/6023ccb96d0371acc01ccb5945980992ca999441)

technophile-04 reviewed (approved):
> Possible solution is to rewrite Address type directly, not through Register. It works, but am I missing something? Do we really need to use Register. Also updated TS version and looks like everything works fine.

Actually it was suggested in abitype docs:  https://abitype.dev/config#overview

also while digging I found this in their [release logs](https://github.com/wevm/abitype/releases): 
![Screenshot 2024-07-16 at 1 43 45 PM](https://github.com/user-attachments/assets/5e9c1336-50f7-461b-8603-a741b2e643f3)

Maybe this was the reason `Register` were not working? 

--- 

But your current approach seems to works nicely, also tried deploying to vercel and works nicely too.

I think with this approach the problem will reappear again in future when viem / wagmi internal abitype version deviates from our `abitype` version? But since we have locked the version, I think we can only encounter it while manually trying to update the viem / wagmi. 

We can merge this though since currently everything works and its required for https://github.com/scaffold-eth/create-eth-extensions/pull/11. Just cc @carletex once.

rin-st commented:
> Maybe this was the reason Register were not working?

Thanks Shiv, yes! 🙏  After update `abitype` to `1.0.5` problem disappeared, so I switched to `Register` again.
There were no errors but for being more confident I also updated `wagmi`/`viem` to latest versions

rin-st commented on packages/nextjs/types/abitype/abi.d.ts:
```diff
@@ -7,15 +7,3 @@ declare module "abitype" {
   }
 }
-
-declare module "viem/node_modules/abitype" {
```
removed this since it worked with different abitype versions (1.0.0 in viem, 1.0.5 in our package.json)

rin-st commented on packages/nextjs/types/abitype/abi.d.ts:
```diff
@@ -7,15 +7,3 @@ declare module "abitype" {
-  }
-}
-
-declare module "wagmi/node_moudles/abitype" {
```
looks like we don't need to declare module for wagmi at all. No abitype in `wagmi`'s deps, and typo here (`node_moudles`) and everything worked fine

technophile-04 replied:
lol it was hiding all along until now!

technophile-04 reviewed (approved):
> Thanks Shiv, yes! 🙏 After update `abitype` to `1.0.5` problem disappeared, so I switched to `Register` again. There were no errors but for being more confident I also updated `wagmi`/`viem` to latest versions

Noicce! Yup also updating wagmi / viem make sense just to be sure.

All in all, I think everything looks great and works fine!! 

- We are using suggested way to update addressType using `Register`
- Deploys to vercel nicely : https://ts-address-type.vercel.app/
- Also we have switched to `moduleResolution: "Bundler"` 

Thanks Rinat!! Merging this!
